### PR TITLE
Prometheus scaler add option to ignore null value

### PR DESCRIPTION
**Pull request by bamboo12366**
_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [N/A] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes https://github.com/kedacore/keda/issues/3065
Relates to https://github.com/kedacore/keda/issues/3064
Relates to https://github.com/kedacore/keda-docs/pull/769 (docs)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ To learn more about our roadmap, we recommend reading [this document](ROADMAP.md
 - **General:** Use more readable timestamps in KEDA Operator logs ([#3066](https://github.com/kedacore/keda/issue/3066))
 - **Selenium Grid Scaler:** Edge active sessions not being properly counted ([#2709](https://github.com/kedacore/keda/issues/2709))
 - **Selenium Grid Scaler:** Max Sessions implementation issue ([#3061](https://github.com/kedacore/keda/issues/3061))
+- **Prometheus Scaler:** Add ignoreNullValues to return error when prometheus return null in values ([#3065](https://github.com/kedacore/keda/issues/3065))
```
Comment (JorTurFer) [inline CHANGELOG.md:46]: Please, move this line between **General:** and **Selenium Grid Scaler:** for sorting the changelog :pray:

**Review by JorTurFer — COMMENTED** [review]
The change looks good in general.  Thanks for this improvement
Some nits to be addressed:
- There is a test failing `TestPrometheusScalerCortexHeader` 
- Open a PR in [keda-docs](https://github.com/kedacore/keda-docs) updating the doc about prometheus scaler (only in the latest version)
- Inline comment in changelog

Apart from that, IDK if it makes sense adding another e2e test to check this new behaviour, WDYT?

File: CHANGELOG.md (hunk 1 above)
Comment (bamboo12366) [reply CHANGELOG.md]: > The change looks good in general. Thanks for this improvement Some nits to be addressed:
> 
> * There is a test failing `TestPrometheusScalerCortexHeader`
> * Open a PR in [keda-docs](https://github.com/kedacore/keda-docs) updating the doc about prometheus scaler (only in the latest version)
> * Inline comment in changelog
> 
> Apart from that, IDK if it makes sense adding another e2e test to check this new behaviour, WDYT?

The changelog and `TestPrometheusScalerCortexHeader` already updated.
I will update the keda-doc these days.
I think it unnecessary to do e2e in this case, cause I already cover this one in test file by mocking the prometheus return data, which will be the same thing in e2e.

**Comment by bamboo12366** [comment]
> The change looks good in general. Thanks for this improvement Some nits to be addressed:
> 
> * There is a test failing `TestPrometheusScalerCortexHeader`
> * Open a PR in [keda-docs](https://github.com/kedacore/keda-docs) updating the doc about prometheus scaler (only in the latest version)
> * Inline comment in changelog
> 
> Apart from that, IDK if it makes sense adding another e2e test to check this new behaviour, WDYT?



> The change looks good in general. Thanks for this improvement Some nits to be addressed:
> 
> * There is a test failing `TestPrometheusScalerCortexHeader`
> * Open a PR in [keda-docs](https://github.com/kedacore/keda-docs) updating the doc about prometheus scaler (only in the latest version)
> * Inline comment in changelog
> 
> Apart from that, IDK if it makes sense adding another e2e test to check this new behaviour, WDYT?

The changelog and TestPrometheusScalerCortexHeader already updated.
I will update the keda-doc these days.
I think it unnecessary to do e2e in this case, cause I already cover this one in test file by mocking the prometheus return data, which will be the same thing in e2e.

File: CHANGELOG.md (hunk 1 above)
Comment (bamboo12366) [reply CHANGELOG.md]: updated

**Comment by bamboo12366** [comment]
The PR related to keda-doc: https://github.com/kedacore/keda-docs/pull/769

File: pkg/scalers/prometheus_scaler.go [hunk 2]
```diff
@@ -134,6 +144,16 @@ func parsePrometheusMetadata(config *ScalerConfig) (meta *prometheusMetadata, er
 		meta.cortexOrgID = val
 	}
 
+	meta.ignoreNullValues = defaultIgnoreNullValues
+	if val, ok := config.TriggerMetadata[ignoreNullValues]; ok && val != "" {
+		if val == "disable" {
```
Comment (JorTurFer) [inline pkg/scalers/prometheus_scaler.go:149]: Why don't you just use `true` and `false`? There are other cases with `true/false` and in this case I'd say that the meaning doesn't change

Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go:149]: Either true/false or enable/disable is ok for me. The enable/disable feels more legit to me when I wrote the code hhh.

Comment (tomkerkhove) [reply pkg/scalers/prometheus_scaler.go:149]: Let's use true/false to align with other scalers

File: pkg/scalers/prometheus_scaler.go [hunk 3]
```diff
@@ -223,14 +243,20 @@ func (s *prometheusScaler) ExecutePromQuery(ctx context.Context) (float64, error
 
 	// allow for zero element or single element result sets
 	if len(result.Data.Result) == 0 {
-		return 0, nil
+		if s.metadata.ignoreNullValues {
+			return 0, nil
+		}
+		return -1, fmt.Errorf("prometheus metrics %s target maybe lost, the result is empty", s.metadata.metricName)
```
Comment (tomkerkhove) [inline pkg/scalers/prometheus_scaler.go:249]: I think this is more accurate, if I'm not mistaken?
```suggestion
		return -1, fmt.Errorf("prometheus metrics %s target may be lost, the result is empty", s.metadata.metricName)
```

File: pkg/scalers/prometheus_scaler.go [hunk 4]
```diff
@@ -223,14 +243,20 @@ func (s *prometheusScaler) ExecutePromQuery(ctx context.Context) (float64, error
...
 		return -1, fmt.Errorf("prometheus query %s returned multiple elements", s.metadata.query)
 	}
 
 	valueLen := len(result.Data.Result[0].Value)
 	if valueLen == 0 {
-		return 0, nil
+		if s.metadata.ignoreNullValues {
+			return 0, nil
+		}
+		return -1, fmt.Errorf("prometheus metrics %s target maybe lost, the value list is empty", s.metadata.metricName)
```
Comment (tomkerkhove) [inline pkg/scalers/prometheus_scaler.go:259]:
```suggestion
		return -1, fmt.Errorf("prometheus metrics %s target may be lost, the value list is empty", s.metadata.metricName)
```

File: pkg/scalers/prometheus_scaler.go (hunk 2 above)
Comment (tomkerkhove) [reply pkg/scalers/prometheus_scaler.go:149]: Would that be possible for you?

**Review by tomkerkhove — CHANGES_REQUESTED** [review]

File: pkg/scalers/prometheus_scaler.go (hunk 2 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: No problem, updated to true/false already

File: pkg/scalers/prometheus_scaler.go (hunk 3 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: update

File: pkg/scalers/prometheus_scaler.go (hunk 4 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: update

File: pkg/scalers/prometheus_scaler.go [hunk 5]
```diff
@@ -134,6 +144,18 @@ func parsePrometheusMetadata(config *ScalerConfig) (meta *prometheusMetadata, er
...
+	if val, ok := config.TriggerMetadata[ignoreNullValues]; ok && val != "" {
+		ignoreNullValues, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, fmt.Errorf("err incorrect value for ignoreNullValues given: %s, " +
+				"please using ture or false", val)
+		}
+		meta.ignoreNullValues = ignoreNullValues
+	} else {
+		meta.ignoreNullValues = defaultIgnoreNullValues
+	}
```
Comment (JorTurFer) [inline pkg/scalers/prometheus_scaler.go:157]: this else if not needed because the default is set in line 147

**Review by JorTurFer — COMMENTED** [review]
LGTM, only 1 thing inline
Thanks for the improvement! ❤️

File: pkg/scalers/prometheus_scaler.go [hunk 6]
```diff
@@ -134,6 +144,18 @@ func parsePrometheusMetadata(config *ScalerConfig) (meta *prometheusMetadata, er
 		meta.cortexOrgID = val
 	}
 
+	meta.ignoreNullValues = defaultIgnoreNullValues
+	if val, ok := config.TriggerMetadata[ignoreNullValues]; ok && val != "" {
+		ignoreNullValues, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, fmt.Errorf("err incorrect value for ignoreNullValues given: %s, " +
+				"please using ture or false", val)
```
Comment (zroubalik) [inline pkg/scalers/prometheus_scaler.go:152]:
```suggestion
				"please use true or false", val)
```

**Review by zroubalik — COMMENTED** [review]
Looking good, just a few nits.

@NaveLevi would be nice if you can take a look on this

File: pkg/scalers/prometheus_scaler.go (hunk 5 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: sorry, my bad

File: pkg/scalers/prometheus_scaler.go (hunk 6 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: updated, thx you a lot

File: pkg/scalers/prometheus_scaler.go [hunk 7]
```diff
@@ -134,6 +144,16 @@ func parsePrometheusMetadata(config *ScalerConfig) (meta *prometheusMetadata, er
 		meta.cortexOrgID = val
 	}
 
+	meta.ignoreNullValues = defaultIgnoreNullValues
+	if val, ok := config.TriggerMetadata[ignoreNullValues]; ok && val != "" {
+		ignoreNullValues, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, fmt.Errorf("err incorrect value for ignoreNullValues given: %s, " +
+				"please use ture or false", val)
```
Comment (zroubalik) [inline pkg/scalers/prometheus_scaler.go:152]:
```suggestion
				"please use true or false", val)
```
@bamboo12366 there is still typo

**Comment by zroubalik** [comment]
Static check fails as well: https://github.com/kedacore/keda/runs/6652557343?check_suite_focus=true#step:6:84

File: pkg/scalers/prometheus_scaler.go (hunk 7 above)
Comment (bamboo12366) [reply pkg/scalers/prometheus_scaler.go]: sorry...

**Review by JorTurFer — APPROVED** [review]
LGTM! ❤️

**Review by tomkerkhove — APPROVED** [review]

**Pull request merged**